### PR TITLE
Api Audit and Fixing Up Effect Checks

### PR DESCRIPTION
**Pull request by gilbo**
This PR will collect all of my work to make sure the new effect checks are working.

This PR will not yet try to deprecate or strip out the old effect system.

This PR will also not yet try to replace the old bounds check with the new effect-checking machinery.  (Another PR should replace the bounds checking / assert checking, and then yet another PR can strip out the old effect checking stuff entirely)

### CODE REVIEW

File: src/exo/new_eff.py [hunk 1]
```diff
@@ -1528,3 +1532,36 @@ def Check_Bounds(proc, alloc_stmt, block):
     if not is_ok:
         raise SchedulingError(
             f"The buffer {alloc_stmt.name} is accessed out-of-bounds")
+
+def Check_IsDeadAfter(proc, stmts, bufname, ndim):
```
Comment (yamaguchi1024) [inline src/exo/new_eff.py:1536]: Does this function basically replace the dependency analysis for free variables checks?

File: src/exo/pyparser.py [hunk 2]
```diff
@@ -393,8 +393,6 @@ def parse_num_type(self, node, is_arg = False):
                     if isinstance(node.slice, pyast.Tuple):
                         dims = node.slice.elts
                     else:
-                        assert isinstance(node.slice,
```
Comment (yamaguchi1024) [inline src/exo/pyparser.py:396]: Why did you delete this assert?

File: src/exo/new_eff.py (hunk 1 above)
Comment (gilbo) [reply src/exo/new_eff.py:1536]: See the following test: `test_data_reuse_loop_fail()`

File: src/exo/LoopIR_scheduling.py [hunk 3]
```diff
@@ -767,103 +768,102 @@ def __init__(self, proc, stmt):
         self.orig_proc = proc
         self.win_stmt  = stmt
 
+        print(proc)
```
[the remark this answers is not in the document]
Comment (gilbo) [reply src/exo/LoopIR_scheduling.py]: yes, I assume so; it's gone now.

File: src/exo/pyparser.py (hunk 2 above)
Comment (gilbo) [reply src/exo/pyparser.py:396]: I believe I relaxed the constraint that dependent sizes are simple (i.e. only a name or constant). It seems we already have all the infrastructure to support expressions in tensor types but never removed this check. If you're concerned this might break something, maybe we can add some tests to increase our confidence?

**Comment by yamaguchi1024** [comment]
@gilbo Should this be closed then? Are these commits contained in the merged PR?

**Comment by gilbo** [comment]
No. I need to work on this next. It’s another massive PR to fix the underlying analysis

File: tests/golden/test_schedules/test_expand_dim5.txt [hunk 4]
```diff
@@ -1,4 +1,4 @@
 def foo(n: size, x: i8 @ DRAM):
-    a: i8[n] @ DRAM
```
Comment (yamaguchi1024) [inline tests/golden/test_schedules/test_expand_dim5.txt:2]: @gilbo Why does it need to be changed?

File: src/exo/platforms/gemmini.py [hunk 5]
```diff
@@ -942,7 +946,7 @@ def do_st_acc_i8(
 st_acc_i8_s2_v2 = st_acc_i8_s2_v2.reorder_stmts('src_tmp = _', 'ConfigStore.scale = _')
 st_acc_i8_s2_v2 = st_acc_i8_s2_v2.reorder_stmts('src_tmp : _', 'ConfigStore.scale = _')
 st_acc_i8_s2_v2 = st_acc_i8_s2_v2.fission_after('ConfigStore.scale = _', n_lifts=2)
-st_acc_i8_s2_v2 = st_acc_i8_s2_v2.configwrite_after('ConfigStore.scale = _', ConfigStore, 'dst_stride', 'stride(dst, 2)')
+st_acc_i8_s2_v2 = st_acc_i8_s2_v2.configwrite_after('ConfigStore.scale = _', ConfigStore, 'dst_stride', 'stride(dst, 1)')
```
Comment (yamaguchi1024) [inline src/exo/platforms/gemmini.py:949]: What's happening??

Comment (yamaguchi1024) [reply src/exo/platforms/gemmini.py:949]: We should create a stride typecheck test

File: src/exo/LoopIR_scheduling.py [hunk 6]
```diff
@@ -1802,10 +1911,14 @@ def result(self):
         return self.proc
 
     def alloc_check(self, pre, post):
-        if not _is_alloc_free(pre, post):
-            raise SchedulingError("Will not fission here, because "
-                                  "an allocation might be buried "
-                                  "in a different scope than some use-site")
+        pass
+        #pre_allocs  = { s.name for s in pre if isinstance(s,LoopIR.Alloc) }
```
Comment (yamaguchi1024) [inline src/exo/LoopIR_scheduling.py:1915]: Why is this commented out?

File: src/exo/LoopIR_scheduling.py [hunk 7]
```diff
@@ -3388,30 +3513,37 @@ def __init__(self, proc, alloc_site, bounds):
...
+        return new_stmts
 
 # --------------------------------------------------------------------------- #
 # --------------------------------------------------------------------------- #
 # The Passes to export
 
 class Schedules:
+    DoRenameBuf = _DoRenameBuf
+    DoAssignNewVarBefore = _DoAssignNewVarBefore
+    DoDeleteDeadCode = _DoDeleteDeadCode
```
Comment (yamaguchi1024) [inline src/exo/LoopIR_scheduling.py:3546]: We should go through the motivation of these primitives together and also write tests

File: src/exo/parse_fragment.py [hunk 8]
```diff
@@ -174,6 +174,9 @@ def __init__(self, pat, proc, stmt, scope):
         else:
             self.srcinfo = stmt.srcinfo
 
+            assert scope == "before", ("Non-before scope for parsing "
```
Comment (SamirDroubi) [inline src/exo/parse_fragment.py:177]: Change merged in #277

File: src/exo/LoopIR_scheduling.py [hunk 9]
```diff
@@ -3261,6 +3385,7 @@ def __init__(self, proc, new_name, memory, expr):
         proc = InferEffects(proc).result()
 
         super().__init__(proc)
+        Check_Aliasing(self.proc)
```
Comment (SamirDroubi) [inline src/exo/LoopIR_scheduling.py:3388]: Changes to `_DoStageWindow` were applied in #307

File: tests/test_schedules.py [hunk 10]
```diff
@@ -495,6 +660,17 @@ def bar(n: size, A: i8[n]):
     bar = bar.lift_alloc('tmp_a : _', n_lifts=1)
     assert str(bar) == golden
 
+def test_lift_alloc_fv_error():
```
Comment (SamirDroubi) [inline tests/test_schedules.py:663]: Similar test already exists

File: tests/golden/test_schedules/test_inline_window.txt [hunk 11]
```diff
@@ -4,5 +4,5 @@ def foo(n: size, m: size, x: R[n, m] @ DRAM):
     for i in seq(0, n - 4):
         for j in seq(0, m - 4):
             a: R @ DRAM
-            a = x[i, j] * x[2 + i, 1 + i]
-            x[2 + i, 1 + i] = a + x[i + 1, j + 1]
+            a = x[i, j] * x[i + 2, j + 1]
+            x[i + 2, j + 1] = a + x[i + 1, j + 1]
```
Comment (SamirDroubi) [inline tests/golden/test_schedules/test_inline_window.txt:8]: Changes were applied in some commit

File: tests/test_schedules.py [hunk 12]
```diff
@@ -625,10 +801,10 @@ def bar(n: size, src: [R][n], dst: [R][n]):
                 dst[i] = src[i] + src[i + 1]
 
     @proc
-    def foo(x: R[50, 2]):
```
Comment (SamirDroubi) [inline tests/test_schedules.py:628]: Changes applied in #307

File: src/exo/LoopIR_unification.py [hunk 13]
```diff
@@ -48,6 +49,7 @@ def __init__(self, proc, subproc, stmt_block):
         self.live_vars      = ChainMap()
 
         super().__init__(proc)
+        Check_Aliasing(self.proc)
```
Comment (SamirDroubi) [inline src/exo/LoopIR_unification.py:52]: Changes were applied in #307

File: src/exo/LoopIR_scheduling.py [hunk 14]
```diff
@@ -1192,13 +1264,13 @@ def __init__(self, proc, if_stmt, n_lifts):
         assert is_pos_int(n_lifts)
 
         self.target = if_stmt
-        self.loop_deps = vars_in_expr(if_stmt.cond)
```
Comment (SamirDroubi) [inline src/exo/LoopIR_scheduling.py:1195]: `_DoLiftIf` is deprecated now

File: src/exo/LoopIR_scheduling.py [hunk 15]
```diff
@@ -1737,26 +1864,8 @@ def _FV(stmts):
...
-        if isinstance(e, LoopIR.Read):
-            self.vars.add(e.name)
-
-        super().do_e(e)
-
-def vars_in_expr(expr):
-    return _VarsInExpr(expr).result()
-
 
+# remove when _DoDoubleFission is removed
```
Comment (SamirDroubi) [inline src/exo/LoopIR_scheduling.py:1868]: Cannot be deprecated yet, still used in `_FissionLoops`

File: tests/golden/test_schedules/test_unify4.txt [hunk 16]
```diff
@@ -1,2 +1,2 @@
-def foo(x: R[50, 2] @ DRAM):
```
Comment (SamirDroubi) [inline tests/golden/test_schedules/test_unify4.txt:1]: Changes were applied in #307
